include/deal.II/rol/vector_adaptor.h

```diff
@@ -0,0 +1,482 @@
+      Teuchos::dyn_cast< const VectorAdaptor>(rol_vector);
+
+    return (*vector_ptr) * (*vector_adaptor.getVector());
+  }
+
+
+
+  template<typename VectorType>
+  typename VectorType::value_type
+  VectorAdaptor<VectorType>::norm() const
+  {
+    return internal::NumberType<value_type>::value(vector_ptr->l2_norm());
+  }
+
+
+
+  template<typename VectorType>
+  Teuchos::RCP<ROL::Vector<typename VectorType::value_type> >
+  VectorAdaptor<VectorType>::clone() const
+  {
+    Teuchos::RCP< VectorType> vec_ptr = Teuchos::rcp (new VectorType);
+    (*vec_ptr) = (*vector_ptr);
+
+    return Teuchos::rcp (new VectorAdaptor (vec_ptr));
+  }
+
+
+
+  template<typename VectorType>
+  Teuchos::RCP<ROL::Vector<typename VectorType::value_type> >
+  VectorAdaptor<VectorType>::basis (const int i) const
+  {
+    Teuchos::RCP<VectorType> vec_ptr = Teuchos::rcp (new VectorType);
+
+    // Zero all the entries in dealii vector.
+    vec_ptr->reinit(*vector_ptr, false);
+
+    if (vector_ptr->locally_owned_elements().is_element(i))
+      vec_ptr->operator[](i) = 1.;
+
+    vec_ptr->compress(VectorOperation::insert);
+
+    Teuchos::RCP<VectorAdaptor> e = Teuchos::rcp (new VectorAdaptor(vec_ptr));
+
+    return e;
+  }
+
+
+
+  template<typename VectorType>
+  void
+  VectorAdaptor<VectorType>::
+  applyUnary (const ROL::Elementwise::UnaryFunction<value_type> &f)
+  {
+    for (typename VectorType::iterator
+         iterator  = vector_ptr->begin();
+         iterator != vector_ptr->end();
+         iterator++)
+      *iterator = f.apply(*iterator);
+
+    vector_ptr->compress (VectorOperation::insert);
+  }
+
+
+
+  template<typename VectorType>
+  void
+  VectorAdaptor<VectorType>::
+  applyBinary (const ROL::Elementwise::UnaryFunction<value_type> &f,
+               const ROL::Vector<value_type>                     &rol_vector)
+  {
+    Assert (this->dimension() == rol_vector.dimension(),
+            ExcDimensionMismatch(this->dimension(), rol_vector.dimension()));
+
+    const VectorAdaptor &vector_adaptor =
+      Teuchos::dyn_cast<const VectorAdaptor>(rol_vector);
+
+    const VectorType &given_rol_vector = *(vector_adaptor.getVector());
+
+    for (typename VectorType::iterator
+         l_iterator  = vector_ptr->begin(), r_iterator  = given_rol_vector.begin();
+         l_iterator != vector_ptr->end() && r_iterator != given_rol_vector.end();
+         l_iterator++,                      r_iterator++)
+      *l_iterator = f.apply(*l_iterator, *r_iterator);
+
+    vector_ptr->compress (VectorOperation::insert);
+  }
+
+
+
+  template<typename VectorType>
+  typename VectorType::value_type
+  VectorAdaptor<VectorType>::
+  reduce (const ROL::Elementwise::ReductionOp<value_type> &r) const
+  {
+    typename VectorType::value_type result = r.initialValue();
+
+    for (typename VectorType::iterator
+         iterator  = vector_ptr->begin();
+         iterator != vector_ptr->end();
+         iterator++)
+      r.reduce(*iterator, result);
```

Comment:
does ROL perform the parallel reduction between processors itself? If so, maybe add this as a comment before the `return` statement.